File: components-registry-service-light-client/src/main/java/org/octopusden/octopus/components/registry/light/client/ComponentsRegistryServiceErrorDecoder.java
package org.octopusden.octopus.components.registry.light.client;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.octopusden.octopus.components.registry.light.client.dto.ErrorResponse;
import org.octopusden.octopus.components.registry.light.client.exceptions.NotFoundException;
import org.octopusden.octopus.components.registry.light.client.exceptions.RepositoryNotPreparedException;
import feign.Response;
import feign.codec.ErrorDecoder;

import java.io.IOException;
import java.io.InputStream;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class ComponentsRegistryServiceErrorDecoder extends ErrorDecoder.Default {

    private final ObjectMapper objectMapper;

    private static final Map<Integer, Function<ErrorResponse, Exception>> ERROR_RESPONSE_CODES = new HashMap<>();

    static {
        ERROR_RESPONSE_CODES.put(404, (ErrorResponse errorResponse) -> new NotFoundException(errorResponse.getErrorMessage()));
        ERROR_RESPONSE_CODES.put(425, (ErrorResponse errorResponse) -> new RepositoryNotPreparedException(errorResponse.getErrorMessage()));
    }

    public ComponentsRegistryServiceErrorDecoder(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    @Override
    public Exception decode(String methodKey, Response response) {
        if (response != null) {
            ErrorResponse errorResponse = getErrorResponse(response);
            if (errorResponse != null) {
                int status = response.status();
                Function<ErrorResponse, Exception> errorHandler = ERROR_RESPONSE_CODES.get(status);
                if (errorHandler != null) {
                    return errorHandler.apply(errorResponse);
                }
            }
        }
        return super.decode(methodKey, response);
    }

    private ErrorResponse getErrorResponse(Response response) {
        Collection<String> contentTypes = response.headers().getOrDefault("content-type", Collections.emptyList());
        if (contentTypes != null) {
            for (String contentType : contentTypes) {
                if (contentType.contains("application/json")) {
                    try {
                        if (response.body() != null) {
                            try (InputStream inputStream = response.body().asInputStream()) {
                                return objectMapper.readValue(inputStream, ErrorResponse.class);
                            }
                        }
                    } catch (IOException e) {
                        return null;
                    }
                }
            }
        }
        return null;
    }
}

File: components-registry-service-client/src/main/kotlin/org/octopusden/octopus/components/registry/client/ComponentsRegistryServiceErrorDecoder.kt
package org.octopusden.octopus.components.registry.client

import com.fasterxml.jackson.databind.ObjectMapper
import org.octopusden.octopus.components.registry.core.dto.ErrorResponse
import org.octopusden.octopus.components.registry.core.exceptions.NotFoundException
import org.octopusden.octopus.components.registry.core.exceptions.RepositoryNotPreparedException
import feign.Response
import feign.RetryableException
import feign.codec.ErrorDecoder

class ComponentsRegistryServiceErrorDecoder(val objectMapper: ObjectMapper) : ErrorDecoder.Default() {

    override fun decode(methodKey: String?, response: Response?): Exception {
        if (isRetryable(response)) {
            // Let super parse Retry-After header first; reuse the RetryableException it builds
            // (which carries the parsed retry timestamp). If super returns a plain FeignException
            // (no Retry-After header present), synthesize our own with null retryAfter.
            val superException = super.decode(methodKey, response)
            if (superException is RetryableException) {
                return superException
            }
            return RetryableException(
                response!!.status(),
                response.reason() ?: "Service Unavailable",
                response.request().httpMethod(),
                null as java.util.Date?,
                response.request()
            )
        }
        return getErrorResponse(response)
                ?.let {
                    val status = response?.status()!!
                    errorResponseCodes.getOrDefault(status) { super.decode(methodKey, response) }
                            .invoke(it)
                } ?: super.decode(methodKey, response)
    }

    private fun getErrorResponse(response: Response?): ErrorResponse? {
        return response?.let { res ->
            res.headers()["content-type"]
                    ?.find { it.contains("application/json") }
                    ?.let {
                        try {
                            res.body()
                                    ?.asInputStream()
                                    .use { inputStream -> objectMapper.readValue(inputStream, ErrorResponse::class.java) }
                        } catch (e: Exception) {
                            null
                        }
                    }
        }
    }

    companion object {
        private val errorResponseCodes: Map<Int, (ErrorResponse) -> Exception> = mapOf(
                404 to { errorResponse: ErrorResponse -> NotFoundException(errorResponse.errorMessage) },
                425 to { errorResponse: ErrorResponse -> RepositoryNotPreparedException(errorResponse.errorMessage) }
        )
        private val retryableStatusCodes = setOf(503)
    }

    private fun isRetryable(response: Response?) = response?.status() in retryableStatusCodes
}
